src/lem/circuit.rs

```diff
@@ -0,0 +1,889 @@
+
+                        let new_path = path.push_tag(tag);
+                        stack.push((op, concrete_path_and_has_match, new_path));
+                    }
+
+                    // Now we need to enforce that at least one path was taken. We do that by constraining
+                    // that the sum of the previously collected `Boolean`s is one
+
+                    enforce_selector_with_premise(
+                        &mut cs.namespace(|| format!("{path}.enforce_selector_with_premise")),
+                        &concrete_path,
+                        &concrete_path_vec,
+                    )
+                    .with_context(|| " couldn't constrain `enforce_selector_with_premise`")?;
+                }
+                LEMOP::Seq(ops) => {
+                    stack.extend(
+                        ops.iter()
+                            .rev()
+                            .map(|op| (op, concrete_path.clone(), path.clone())),
+                    );
+                }
+                LEMOP::Return(outputs) => {
+                    for (i, output) in outputs.iter().enumerate() {
+                        let Some(allocated_ptr) = allocated_ptrs.get(output.name()) else {
+                            bail!("{output} not allocated")
+                        };
+
+                        allocated_ptr
+                            .implies_ptr_equal(
+                                &mut cs.namespace(|| {
+                                    format!("{path}.implies_ptr_equal {output} (output {i})")
+                                }),
+                                &concrete_path,
+                                &preallocated_outputs[i],
+                            )
+                            .with_context(|| "couldn't constrain `implies_ptr_equal`")?;
+                    }
+                }
+                _ => todo!(),
+            }
+        }
+        Ok(())
+    }
+
+    /// Computes the number of constraints that `synthesize` should create. It's
+    /// also an explicit way to document and attest how the number of constraints
+    /// grow.
+    pub fn num_constraints(&self, slots_info: &SlotsInfo) -> usize {
+        let mut num_constraints = 0;
+
+        // fixed cost for each slot
+        for (_, slot_type) in &slots_info.slots {
+            match slot_type {
+                SlotType::Hash2 => {
+                    num_constraints += 289;
+                }
+                SlotType::Hash3 => {
+                    num_constraints += 337;
+                }
+                SlotType::Hash4 => {
+                    num_constraints += 388;
+                }
+            }
+        }
+
+        let mut stack = vec![(&self.lem_op, false)];
+        while let Some((op, nested)) = stack.pop() {
+            match op {
+                LEMOP::Null(..) => {
+                    // constrain tag and hash
+                    num_constraints += 2;
```

Comment:
It will be better not to hardcode all these values, but instead compute them, eventually.